Enforce limits on refinement levels and resolution during AMR

kidder opened the pull request:
The default is to limit these to the limits required by the implementation details of ElementId and Spectral.

Fixes https://github.com/sxs-collaboration/spectre/issues/5797

## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
In input files, the you will need to specify a new sub-option to the Amr Policies that specifies the
Limits with sub-options RefinementLevel and NumGridPoints.  The suggested value for these two options is Auto unless you want to limit refinement further.
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

kidder committed:
Enforce limits on refinement levels and resolution during AMR

The default is to limit these to the limits required by the
implementation details of ElementId and Spectral.

kidder commented:
I'll fix the clang-tidy issue when I rebase...

nilsvu commented on src/ParallelAlgorithms/Amr/Policies/EnforcePolicies.cpp:
```diff
@@ -13,17 +16,43 @@
+      gsl::at(*amr_decision, d) = Flag::DoNothing;
+    }
+    if (gsl::at(*amr_decision, d) == Flag::Split and
+        gsl::at(levels, d) == limits.maximum_refinement_level()) {
```
Better make this `>=`? The initial refinement might exceed this based on user input in the domain creator.

nilsvu commented on tests/Unit/ParallelAlgorithms/Amr/Policies/Test_Policies.cpp:
```diff
@@ -10,31 +10,54 @@
+      INFO("specified limits");
+      std::ostringstream creation_string;
+      creation_string << "Isotropy: " << isotropy << "\n";
+      creation_string << "RefinementLimits:\n"
+                      << "  MinimumRefinementLevel: 1\n"
+                      << "  MaximumRefinementLevel: 5\n"
+                      << "  MinimumResolution: 3\n"
+                      << "  MaximumResolution: 7\n";
```
I think you can reduce the verbosity of these options to:

```
Limits:
  RefinementLevel: [0, 8]  # or Auto
  NumGridPoints: [2, 12]  # or Auto
```

(parse the two options as `Options::Auto<std::pair>` and add some parse errors in the constructor)

This accomplishes the following:

- The name `RefinementLimits` is redundant with `Min/MaxRefinementLevel`, but doesn't fit `Min/MaxResolution`, so better name it just `Limits` (that's clear enough in the AMR context).
- The name `Resolution` doesn't fit together with the domain creators, which specify `InitialGridPoints`. To make it clear that the two options refer to the same thing, name it `NumGridPoints`. (I understand that p-refinement is called "resolution" in some places and I don't have better naming for that at the moment, but I think making this option consistent with the domain creator is important.)
- I find the `Minimum` and `Maximum` naming overly verbose, and a simple pair of numbers quite clear and easier to read.
- This is also short enough that you don't need `Auto` for the parent option (`Limits`), but just for each of the two pairs. So I'd avoid hiding the two options behind a `Limits: Auto`. Instead you can set `Auto` for either h-refinement, or p-refinement, or both.

nilsvu commented on tests/InputFiles/ExampleExecutables/RandomAmr1D.yaml:
```diff
@@ -25,6 +25,7 @@ Amr:
         MaximumRefinementLevel: 8
   Policies:
     Isotropy: Anisotropic
+    RefinementLimits: Auto
```
Set max refinement level to 8, like the option above? Same for the ExportCoordinates input file.

Also, I think you can delete the max refinement level from the `Random` criterion in this PR, should be very quick to do.

nilsvu reviewed (changes requested):
Nice! I'll just close https://github.com/sxs-collaboration/spectre/pull/5798 then, or do you want to merge that quickly to avoid CI failures while you rebase this PR?

Feel free to squash changes directly for the changes below.

kidder committed:
Remove MaximumRefinementLevel from amr::Criteria::Random

The limits on refinement levels in now enforced by the amr::Policies.

kidder commented:
@nilsvu I incorporated all of your suggestions and pushed a new version as I didn't think a fixup commit would be any easier to look at.  Also added a commit removing the max level from amr::Criteria::Random

nilsvu commented:
Test just needs a `SPECTRE_DEBUG`

kidder commented:
> Test just needs a `SPECTRE_DEBUG`

done
